docs/TPPDialect.md

```diff
@@ -0,0 +1,355 @@
+
+TPP operations can carry attributes, their operands can be chained and grouped, forming trees and graphs, which can be mathed in different ways.
+
+Once lowered to `xsmm` dialect, the operations are selected and any further lowering won't change the TPP semantics, only add `xsmm` boilerplate and infrastruture to execute the chosen strategies.
+
+# Basic Rules
+
+## Syntax
+
+All `tensor` operations follow the SSA syntax, operands and return values:
+```mlir
+  // Always out-of-place
+  %C = tpp.add ins(%arg0, %arg1) : (tensor<>, tensor<>) -> tensor<>
+  // %C is here treated as initialization, accumulation can still happen on same buffer
+  %M = tpp.matmul ins(%A, %B, %C) : (tensor<>, tensor<>, tensor<>) -> tensor<>
+```
+
+All `memref` operations follow the `linalg` syntax of `ins` and `outs`.
+The `outs` values are an indication of where the output is written to.
+These are generated by the bufferization process, from the `tensor` level operations.
+```mlir
+  // Out-of-place ops output on new allocations
+  %buf = memref.alloc : memref<>
+  tpp.add ins(%arg0, %arg1) outs(%buf) : (memref<>, memref<>)
+  // In-of-place ops output on input operands
+  tpp.matmul ins(%A, %B, %C) outs(%C) : (memref<>, memref<>, memref<>)
+```
+
+### Unary
+
+On `tensors`, unary ops have one argument and one return value.
+
+On `memrefs`, unary operations have only `outs` for non-transforms (neg, inv, exp, etc.).
+For transforms (`broadcast`, `reduce`, `transpose`, `pack`, etc), there is one `ins` argument and one `outs` argument.
+The sizes must be the same for element-wise / arith, but will often be different for transforms (except `copy` and `transpose`).
+
+There is no implicit transform semantics.
+When lowering a broadcast (ex. a `linalg.generic` with lower dimension argument and apporpriate affine map), we map to `tpp.broadcast` + the operation.
+
+### Binary
+
+On `tensors`, binary ops have two arguments and one return value.
+
+On `memrefs`, binary operations have both `ins` and `outs`, with always two `ins` and one `outs` arguments.
+On most binary ops, the `outs` can alias `ins` (ex. `a=a+b`), and `ins` can alias themselves (ex. `a=a+a`).
+
+The exception is `matmul`, where the output cannot alias the inputs.
+A binary `matmul` is equivalent to `C = A x B`, with `C` as a new `memref` allocation.
+
+Element wise operations must have the same types, with broadcast done as a separate op, per operand.
+
+GEMM operations must have compatible types (MxNxK, VNNI, etc.).
+
+### Ternary
+
+On `tensors`, ternary ops have three arguments and one return value.
+
+On `memrefs`, ternary operations have both `ins` and `outs`, with always three `ins` and one `outs` arguments.
+The `outs` can alias with any type-compatible `ins` argument (accumulation matrix).
+
+A ternary `matmul` is equivalent to `C = A x B + C`, with `C` as an input and a potential output.
```

Comment:
This can be split into two separate operations `matmul` and `add` so, no need for ternary variant.

Reply:
Good point. We can make these semantics more explicit with ternary.
The third operand could be either directly an output `C` matrix or a scalar value `β`. But as we probably aren't interested in supporting non-zero or non-one `β`. Passing output storage `C` makes more sense.

> If anything, we can probably drop the binary version and force matmul to have a zero initializer as the third element.

I think that's a good idea. We should support only single op format whenever possible.